mewarpx/mewarpx/diags_store/particle_diagnostic.py

```diff
@@ -0,0 +1,155 @@
+
+            species (mepicmi.Species): species in the simulation, if None then
+            uses all particles in the simulation
+
+            data_list (list str): list of attributes to be outputted by he
+            diagnostic, default uses ["position", "momentum", "weighting"]
+
+            write_dir (str): where to write particle diagnostic data
+
+            post_process (bool): generate plots for each diagnostic data
+            directory produced
+
+            plot_data_list (list str): list of data to be plotted for each
+            diagnostic step ("particle_position_x", "particle_position_y",
+            "particle_position_z", "particle_momentum_x", "particle_momentum_y",
+            "particle_momentum_z")
+
+            plot_species (list str): list of species names to be plotted,
+            defaults to all species if not specified. Name variable in
+            mepicmi.Species must be set for each species in the simulation
+
+        """
+        self.name = name
+        self.species = species
+        self.data_list = data_list
+        self.write_dir = write_dir
+
+        self.post_process = post_process
+        self.plot_data_list = plot_data_list
+        self.plot_species = plot_species
+
+        if self.name is None:
+            self.name= "diag"
+        if self.species is None:
+            self.species = mwxrun.simulation.species
+        if self.data_list is None:
+            self.data_list = ["position", "momentum", "weighting"]
+        if self.plot_species is None:
+            self.plot_species = []
+            for specimen in self.species:
+                self.plot_species.append(specimen.name)
+
+        super(ParticleDiagnostic, self).__init__(diag_steps=diag_steps, **kwargs)
+        self.add_particle_diag()
+
+        if self.post_process:
+            if self.plot_data_list is None:
+                raise UserWarning("Warning: post_process is True,"
+                                "but plot_data_list is None!")
+            else:
+                self.add_post_processing_check_callback()
+
+
+    def add_particle_diag(self):
+        particle_diag = picmi.ParticleDiagnostic(
+            name=self.name,
+            period=self.diag_steps,
+            species=self.species,
+            data_list=self.data_list,
+            write_dir=self.write_dir
+        )
+
+        mwxrun.simulation.add_diagnostic(particle_diag)
+
+    def diag_data_path(self):
+        read_location = None
+        if self.write_dir == '.':
+            read_location = os.getcwd()
+            read_location = os.path.join(read_location, "diags")
+        else:
+            read_location = self.write_dir
+
+        return read_location
+
+    def check_for_end_of_sim(self):
+        if mwxrun.get_it() == mwxrun.simulation.max_steps:
+            self.do_post_processing()
+
+    def add_post_processing_check_callback(self):
+        callbacks.installafterstep(self.check_for_end_of_sim)
```

Comment:
One thing Peter asked me to do was to just move this into `__init__` because this extra layer of function calls wasn't necessary.